File: cpp/CommandBasedDrive/src/main/cpp/subsystems/DriveSubsystem.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "subsystems/DriveSubsystem.h"
#include "ctre/phoenixpro/Utils.hpp"
#include <frc/smartdashboard/SmartDashboard.h>
#include <frc/RobotController.h>

using namespace ctre::phoenixpro;

DriveSubsystem::DriveSubsystem()
{
    /* Initialize all the devices */
    InitializeLeftDriveTalonFX(m_leftLeader.GetConfigurator());
    InitializeLeftDriveTalonFX(m_leftFollower.GetConfigurator());
    InitializeRightDriveTalonFX(m_rightLeader.GetConfigurator());
    InitializeRightDriveTalonFX(m_rightFollower.GetConfigurator());
    InitializePigeon2(m_pigeon2.GetConfigurator());

    /* Set followers to follow leader */
    m_leftFollower.SetControl(controls::Follower{m_leftLeader.GetDeviceID(), false});
    m_rightFollower.SetControl(controls::Follower{m_rightLeader.GetDeviceID(), false});

    /* Make sure all critical signals are synchronized */
    /* Setting all these signals to 100hz means they get sent at the same time if they're all on a CANivore */
    m_pigeon2.GetYaw().SetUpdateFrequency(100_Hz);
    m_leftLeader.GetPosition().SetUpdateFrequency(100_Hz);
    m_rightLeader.GetPosition().SetUpdateFrequency(100_Hz);

    /* Set the update frequency of the main requests to 0 so updates are sent immediately in the arcadeDrive method */
    m_leftOut.UpdateFreqHz = 0_Hz;
    m_rightOut.UpdateFreqHz = 0_Hz;
    
    /* Currently in simulation, we do not support FOC, so disable it while simulating */
    if (ctre::phoenixpro::IsSimulation())
    {
        m_leftOut.EnableFOC = false;
        m_rightOut.EnableFOC = false;
    }

    /*
     * Set the orientation of the simulated TalonFX devices relative to the robot chassis.
     * WPILib expects +V to be forward. Specify orientations to match that behavior.
     */
    /* left TalonFXs are CCW+ */
    m_leftSimState.orientation = sim::ChassisReference::CounterClockwise_Positive;
    m_leftFollowerSimState.orientation = sim::ChassisReference::CounterClockwise_Positive;
    /* right TalonFXs are CW+ */
    m_rightSimState.orientation = sim::ChassisReference::Clockwise_Positive;
    m_rightFollowerSimState.orientation = sim::ChassisReference::Clockwise_Positive;

    /* Publish field pose data to read back from */
    frc::SmartDashboard::PutData("Field", &m_field);
}

void DriveSubsystem::ArcadeDrive(double fwd, double rot)
{
    m_leftOut.Output = fwd + rot;
    m_rightOut.Output = fwd - rot;
    m_leftLeader.SetControl(m_leftOut);
    m_rightLeader.SetControl(m_rightOut);
}

void DriveSubsystem::InitializeLeftDriveTalonFX(ctre::phoenixpro::configs::TalonFXConfigurator &cfg)
{
    configs::TalonFXConfiguration toApply{};

    /* User can change configs if they want, or leave this blank for factory-default */
    toApply.MotorOutput.Inverted = false;

    cfg.Apply(toApply);

    /* And initialize position to 0 */
    cfg.SetRotorPosition(0_tr);
}
void DriveSubsystem::InitializeRightDriveTalonFX(ctre::phoenixpro::configs::TalonFXConfigurator &cfg)
{
    configs::TalonFXConfiguration toApply{};

    /* User can change configs if they want, or leave this blank for factory-default */
    toApply.MotorOutput.Inverted = true;

    cfg.Apply(toApply);

    /* And initialize position to 0 */
    cfg.SetRotorPosition(0_tr);
}
void DriveSubsystem::InitializePigeon2(ctre::phoenixpro::configs::Pigeon2Configurator &cfg)
{
    configs::Pigeon2Configuration toApply{};

    /* User can change configs if they want, or leave this blank for factory-default */

    cfg.Apply(toApply);

    /* And initialize yaw to 0 */
    cfg.SetYaw(0_deg);
}

void DriveSubsystem::Periodic()
{
    /*
     * This will get the simulated sensor readings that we set
     * in the previous article while in simulation, but will use
     * real values on the robot itself.
     */
    m_odometry.Update(m_pigeon2.GetRotation2d(),
            rotationsToMeters(m_leftLeader.GetPosition().GetValue()),
            rotationsToMeters(m_rightLeader.GetPosition().GetValue()));
    m_field.SetRobotPose(m_odometry.GetPose());
}

void DriveSubsystem::SimulationPeriodic()
{
    /* Pass the robot battery voltage to the simulated devices */
    m_leftSimState.SetSupplyVoltage(frc::RobotController::GetBatteryVoltage());
    m_rightSimState.SetSupplyVoltage(frc::RobotController::GetBatteryVoltage());
    m_leftFollowerSimState.SetSupplyVoltage(frc::RobotController::GetBatteryVoltage());
    m_rightFollowerSimState.SetSupplyVoltage(frc::RobotController::GetBatteryVoltage());
    m_pigeon2SimState.SetSupplyVoltage(frc::RobotController::GetBatteryVoltage());

    /*
     * CTRE simulation is low-level, so SimState inputs
     * and outputs are not affected by user-level inversion.
     * However, inputs and outputs *are* affected by the mechanical
     * orientation of the device relative to the robot chassis,
     * as specified by the `orientation` field.
     *
     * WPILib expects +V to be forward. We have already configured
     * our orientations to match this behavior.
     */
    m_driveSim.SetInputs(m_leftSimState.GetMotorVoltage(),
                         m_rightSimState.GetMotorVoltage());

    /*
     * Advance the model by 20 ms. Note that if you are running this
     * subsystem in a separate thread or have changed the nominal
     * timestep of TimedRobot, this value needs to match it.
     */
    m_driveSim.Update(20_ms);

    /* Update all of our sensors. */
    m_leftSimState.SetRawRotorPosition(
        metersToRotations(m_driveSim.GetLeftPosition())
    );
    m_leftSimState.SetRotorVelocity(
        metersToRotationsVel(m_driveSim.GetLeftVelocity())
    );
    m_rightSimState.SetRawRotorPosition(
        metersToRotations(m_driveSim.GetRightPosition())
    );
    m_rightSimState.SetRotorVelocity(
        metersToRotationsVel(m_driveSim.GetRightVelocity())
    );
    m_leftFollowerSimState.SetRawRotorPosition(
        metersToRotations(m_driveSim.GetLeftPosition())
    );
    m_leftFollowerSimState.SetRotorVelocity(
        metersToRotationsVel(m_driveSim.GetLeftVelocity())
    );
    m_rightFollowerSimState.SetRawRotorPosition(
        metersToRotations(m_driveSim.GetRightPosition())
    );
    m_rightFollowerSimState.SetRotorVelocity(
        metersToRotationsVel(m_driveSim.GetRightVelocity())
    );
    m_pigeon2SimState.SetRawYaw(m_driveSim.GetHeading().Degrees());
}

units::meter_t DriveSubsystem::rotationsToMeters(units::turn_t rotations)
{
    /* Get circumference of wheel */
    constexpr auto circumference = kWheelRadiusInches * 2 * 3.14159 / 1_tr;
    /* Every rotation of the wheel travels this many inches */
    /* Now apply gear ratio to input rotations */
    auto gearedRotations = rotations / kGearRatio;
    /* And multiply geared rotations by meters per rotation */
    return gearedRotations * circumference;
}

units::turn_t DriveSubsystem::metersToRotations(units::meter_t meters)
{
    /* Get circumference of wheel */
    constexpr auto circumference = kWheelRadiusInches * 2 * 3.14159 / 1_tr;
    /* Every rotation of the wheel travels this many inches */
    /* Now apply wheel rotations to input meters */
    auto wheelRotations = meters / circumference;
    /* And multiply by gear ratio to get rotor rotations */
    return wheelRotations * kGearRatio;
}

units::meters_per_second_t DriveSubsystem::rotationsToMetersVel(units::turns_per_second_t rotations)
{
    /* Get circumference of wheel */
    constexpr auto circumference = kWheelRadiusInches * 2 * 3.14159 / 1_tr;
    /* Every rotation of the wheel travels this many inches */
    /* Now apply gear ratio to input rotations */
    auto gearedRotations = rotations / kGearRatio;
    /* And multiply geared rotations by meters per rotation */
    return gearedRotations * circumference;
}

units::turns_per_second_t DriveSubsystem::metersToRotationsVel(units::meters_per_second_t meters)
{
    /* Get circumference of wheel */
    constexpr auto circumference = kWheelRadiusInches * 2 * 3.14159 / 1_tr;
    /* Every rotation of the wheel travels this many inches */
    /* Now apply wheel rotations to input meters */
    auto wheelRotations = meters / circumference;
    /* And multiply by gear ratio to get rotor rotations */
    return wheelRotations * kGearRatio;
}

File: java/CommandBasedDrive/src/main/java/frc/robot/subsystems/DriveSubsystem.java
package frc.robot.subsystems;

import static edu.wpi.first.units.Units.*;

import com.ctre.phoenix6.BaseStatusSignal;
import com.ctre.phoenix6.StatusSignal;
import com.ctre.phoenix6.Utils;
import com.ctre.phoenix6.configs.Pigeon2Configuration;
import com.ctre.phoenix6.configs.Pigeon2Configurator;
import com.ctre.phoenix6.configs.TalonFXConfiguration;
import com.ctre.phoenix6.configs.TalonFXConfigurator;
import com.ctre.phoenix6.controls.DutyCycleOut;
import com.ctre.phoenix6.controls.Follower;
import com.ctre.phoenix6.hardware.Pigeon2;
import com.ctre.phoenix6.hardware.TalonFX;
import com.ctre.phoenix6.sim.ChassisReference;
import com.ctre.phoenix6.sim.Pigeon2SimState;
import com.ctre.phoenix6.sim.TalonFXSimState;
import com.ctre.phoenix6.signals.InvertedValue;

import edu.wpi.first.math.kinematics.DifferentialDriveOdometry;
import edu.wpi.first.math.system.plant.DCMotor;
import edu.wpi.first.math.util.Units;
import edu.wpi.first.units.AngleUnit;
import edu.wpi.first.units.measure.Angle;
import edu.wpi.first.units.measure.AngularVelocity;
import edu.wpi.first.units.measure.Distance;
import edu.wpi.first.units.measure.LinearVelocity;
import edu.wpi.first.wpilibj.RobotController;
import edu.wpi.first.wpilibj.simulation.DifferentialDrivetrainSim;
import edu.wpi.first.wpilibj.smartdashboard.Field2d;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import static frc.robot.Constants.*;

public class DriveSubsystem extends SubsystemBase {
    private final TalonFX m_leftLeader = new TalonFX(LEFT_LEADER_ID, CANBUS_NAME);
    private final TalonFX m_leftFollower = new TalonFX(LEFT_FOLLOWER_ID, CANBUS_NAME);
    private final TalonFX m_rightLeader = new TalonFX(RIGHT_LEADER_ID, CANBUS_NAME);
    private final TalonFX m_rightFollower = new TalonFX(RIGHT_FOLLOWER_ID, CANBUS_NAME);

    private final TalonFXSimState m_leftSimState = m_leftLeader.getSimState();
    private final TalonFXSimState m_rightSimState = m_rightLeader.getSimState();
    private final TalonFXSimState m_leftFollowerSimState = m_leftFollower.getSimState();
    private final TalonFXSimState m_rightFollowerSimState = m_rightFollower.getSimState();

    private final DutyCycleOut m_leftOut = new DutyCycleOut(0); // Initialize with 0% output
    private final DutyCycleOut m_rightOut = new DutyCycleOut(0); // Initialize with 0% output

    private final Pigeon2 m_pigeon2 = new Pigeon2(PIGEON2_ID, CANBUS_NAME);

    private final Pigeon2SimState m_pigeon2SimState = m_pigeon2.getSimState();

    /*
     * These numbers are an example AndyMark Drivetrain with some additional weight.
     * This is a fairly light robot.
     * Note you can utilize results from robot characterization instead of
     * theoretical numbers.
     * https://docs.wpilib.org/en/stable/docs/software/wpilib-tools/robot-
     * characterization/introduction.html#introduction-to-robot-characterization
     */
    private final double kGearRatio = 10.71;
    private final Distance kWheelRadius = Inches.of(3);

    /* Simulation model of the drivetrain */
    private final DifferentialDrivetrainSim m_driveSim = new DifferentialDrivetrainSim(
        DCMotor.getFalcon500Foc(2), // 2 CIMS on each side of the drivetrain.
        kGearRatio, // Standard AndyMark Gearing reduction.
        2.1, // MOI of 2.1 kg m^2 (from CAD model).
        26.5, // Mass of the robot is 26.5 kg.
        kWheelRadius.in(Meters), // Robot uses 3" radius (6" diameter) wheels.
        0.546, // Distance between wheels is _ meters.

        /*
         * The standard deviations for measurement noise:
         * x and y: 0.001 m
         * heading: 0.001 rad
         * l and r velocity: 0.1 m/s
         * l and r position: 0.005 m
         */
        /* Uncomment the following line to add measurement noise. */
        null // VecBuilder.fill(0.001, 0.001, 0.001, 0.1, 0.1, 0.005, 0.005)
    );

    private final Field2d m_field = new Field2d();
    /*
     * Creating my odometry object. Here,
     * our starting pose is 5 meters along the long end of the field and in the
     * center of the field along the short end, facing forward.
     */
    private final DifferentialDriveOdometry m_odometry = new DifferentialDriveOdometry(
        m_pigeon2.getRotation2d(),
        0, 0
    );

    public DriveSubsystem() {
        /* Initialize all the devices */
        initializeLeftDriveTalonFX(m_leftLeader.getConfigurator());
        initializeLeftDriveTalonFX(m_leftFollower.getConfigurator());
        initializeRightDriveTalonFX(m_rightLeader.getConfigurator());
        initializeRightDriveTalonFX(m_rightFollower.getConfigurator());
        initializePigeon2(m_pigeon2.getConfigurator());

        /* Set followers to follow leader */
        m_leftFollower.setControl(new Follower(m_leftLeader.getDeviceID(), false));
        m_rightFollower.setControl(new Follower(m_rightLeader.getDeviceID(), false));

        /* Make sure all critical signals are synchronized */
        /*
         * Setting all these signals to 100hz means they get sent at the same time if
         * they're all on a CANivore
         */
        BaseStatusSignal.setUpdateFrequencyForAll(100,
            m_leftLeader.getPosition(),
            m_rightLeader.getPosition(),
            m_pigeon2.getYaw());

        /*
         * Set the update frequency of the main requests to 0 so updates are sent
         * immediately in the arcadeDrive method
         */
        m_leftOut.UpdateFreqHz = 0;
        m_rightOut.UpdateFreqHz = 0;

        /*
         * Set the orientation of the simulated TalonFX devices relative to the robot chassis.
         * WPILib expects +V to be forward. Specify orientations to match that behavior.
         */
        /* left TalonFXs are CCW+ */
        m_leftSimState.Orientation = ChassisReference.CounterClockwise_Positive;
        m_leftFollowerSimState.Orientation = ChassisReference.CounterClockwise_Positive;
        /* right TalonFXs are CW+ */
        m_rightSimState.Orientation = ChassisReference.Clockwise_Positive;
        m_rightFollowerSimState.Orientation = ChassisReference.Clockwise_Positive;

        /* Publish field pose data to read back from */
        SmartDashboard.putData("Field", m_field);
    }

    /**
     * Drive the robot using an arcade drive format.
     * <p>
     * <b>This must be called periodically</b> or else the control frames will not
     * get sent
     * out, resulting in the TalonFXs disabling
     * 
     * @param fwd Forward/Reverse output
     * @param rot Left/Right output
     */
    public void arcadeDrive(double fwd, double rot) {
        m_leftOut.Output = fwd + rot;
        m_rightOut.Output = fwd - rot;

        m_leftLeader.setControl(m_leftOut);
        m_rightLeader.setControl(m_rightOut);
    }

    @Override
    public void periodic() {
        /*
         * This will get the simulated sensor readings that we set
         * in the previous article while in simulation, but will use
         * real values on the robot itself.
         */
        m_odometry.update(m_pigeon2.getRotation2d(),
                rotationsToMeters(m_leftLeader.getPosition().getValue()).in(Meters),
                rotationsToMeters(m_rightLeader.getPosition().getValue()).in(Meters));
        m_field.setRobotPose(m_odometry.getPoseMeters());
    }

    @Override
    public void simulationPeriodic() {
        /* Pass the robot battery voltage to the simulated devices */
        m_leftSimState.setSupplyVoltage(RobotController.getBatteryVoltage());
        m_rightSimState.setSupplyVoltage(RobotController.getBatteryVoltage());
        m_leftFollowerSimState.setSupplyVoltage(RobotController.getBatteryVoltage());
        m_rightFollowerSimState.setSupplyVoltage(RobotController.getBatteryVoltage());
        m_pigeon2SimState.setSupplyVoltage(RobotController.getBatteryVoltage());

        /*
         * CTRE simulation is low-level, so SimState inputs
         * and outputs are not affected by user-level inversion.
         * However, inputs and outputs *are* affected by the mechanical
         * orientation of the device relative to the robot chassis,
         * as specified by the `orientation` field.
         *
         * WPILib expects +V to be forward. We have already configured
         * our orientations to match this behavior.
         */
        m_driveSim.setInputs(m_leftSimState.getMotorVoltage(),
                m_rightSimState.getMotorVoltage());

        /*
         * Advance the model by 20 ms. Note that if you are running this
         * subsystem in a separate thread or have changed the nominal
         * timestep of TimedRobot, this value needs to match it.
         */
        m_driveSim.update(0.02);

        /* Update all of our sensors. */
        m_leftSimState.setRawRotorPosition(
            metersToRotations(Meters.of(m_driveSim.getLeftPositionMeters()))
        );
        m_leftSimState.setRotorVelocity(
            // This is OK, since the time base is the same
            metersToRotationsVel(MetersPerSecond.of(m_driveSim.getLeftVelocityMetersPerSecond()))
        );
        m_rightSimState.setRawRotorPosition(
            metersToRotations(Meters.of(m_driveSim.getRightPositionMeters()))
        );
        m_rightSimState.setRotorVelocity(
            // This is OK, since the time base is the same
            metersToRotationsVel(MetersPerSecond.of(m_driveSim.getRightVelocityMetersPerSecond()))
        );
        m_leftFollowerSimState.setRawRotorPosition(
            metersToRotations(Meters.of(m_driveSim.getLeftPositionMeters()))
        );
        m_leftFollowerSimState.setRotorVelocity(
            // This is OK, since the time base is the same
            metersToRotationsVel(MetersPerSecond.of(m_driveSim.getLeftVelocityMetersPerSecond()))
        );
        m_rightFollowerSimState.setRawRotorPosition(
            metersToRotations(Meters.of(m_driveSim.getRightPositionMeters()))
        );
        m_rightFollowerSimState.setRotorVelocity(
            // This is OK, since the time base is the same
            metersToRotationsVel(MetersPerSecond.of(m_driveSim.getRightVelocityMetersPerSecond()))
        );
        m_pigeon2SimState.setRawYaw(m_driveSim.getHeading().getDegrees());
    }

    public StatusSignal<Angle> getYaw() {
        return m_pigeon2.getYaw();
    }

    public StatusSignal<Angle> getLeftPos() {
        return m_leftLeader.getPosition();
    }

    public StatusSignal<Angle> getRightPos() {
        return m_rightLeader.getPosition();
    }

    /**
     * Initialize a left drive TalonFX device from the configurator object
     * 
     * @param cfg Configurator of the TalonFX device
     */
    private void initializeLeftDriveTalonFX(TalonFXConfigurator cfg) {
        var toApply = new TalonFXConfiguration();

        /*
         * User can change configs if they want, or leave this blank for factory-default
         */
        toApply.MotorOutput.Inverted = InvertedValue.CounterClockwise_Positive;

        cfg.apply(toApply);

        /* And initialize position to 0 */
        cfg.setPosition(0);
    }

    /**
     * Initialize a right drive TalonFX device from the configurator object
     * 
     * @param cfg Configurator of the TalonFX device
     */
    private void initializeRightDriveTalonFX(TalonFXConfigurator cfg) {
        var toApply = new TalonFXConfiguration();

        /*
         * User can change configs if they want, or leave this blank for factory-default
         */
        toApply.MotorOutput.Inverted = InvertedValue.Clockwise_Positive;

        cfg.apply(toApply);

        /* And initialize position to 0 */
        cfg.setPosition(0);
    }

    /**
     * Initialize Pigeon2 device from the configurator object
     * 
     * @param cfg Configurator of the Pigeon2 device
     */
    private void initializePigeon2(Pigeon2Configurator cfg) {
        var toApply = new Pigeon2Configuration();

        /*
         * User can change configs if they want, or leave this blank for factory-default
         */

        cfg.apply(toApply);

        /* And initialize yaw to 0 */
        cfg.setYaw(48);
    }

    private Distance rotationsToMeters(Angle rotations) {
        /* Apply gear ratio to input rotations */
        var gearedRadians = rotations.in(Radians) / this.kGearRatio;
        /* Then multiply the wheel radius by radians of rotation to get distance */
        return this.kWheelRadius.times(gearedRadians);
    }

    private Angle metersToRotations(Distance meters) {
        /* Divide the distance by the wheel radius to get radians */
        var wheelRadians = meters.in(Meters) / this.kWheelRadius.in(Meters);
        /* Then multiply by gear ratio to get rotor rotations */
        return Radians.of(wheelRadians * this.kGearRatio);
    }

    private LinearVelocity rotationsToMetersVel(AngularVelocity rotations) {
        /* Apply gear ratio to input rotations */
        var gearedRotations = rotations.divide(this.kGearRatio);
        /* Then multiply the wheel radius by radians of rotation to get distance */
        return this.kWheelRadius.per(Second).times(gearedRotations.in(RadiansPerSecond));
    }

    private AngularVelocity metersToRotationsVel(LinearVelocity meters) {
        /* Divide the distance by the wheel radius to get radians */
        var wheelRadians = meters.in(MetersPerSecond) / this.kWheelRadius.in(Meters);
        /* Then multiply by gear ratio to get rotor rotations */
        return RadiansPerSecond.of(wheelRadians * this.kGearRatio);
    }
}
